WalletWasabi/WabiSabi/Client/CredentialDependencies/CredentialEdgeSet.cs

```diff
@@ -0,0 +1,253 @@
+
+			if (value > 0)
+			{
+				if (RemainingOutDegree(edge.From) == 1)
+				{
+					// This is the final out edge for the node edge.From
+					if (Balance(edge.From) - (long)edge.Value > 0)
+					{
+						throw new InvalidOperationException($"Can't add final out edge without discharging positive value (edge value {edge.Value} but node balance is {Balance(edge.From)}).");
+					}
+
+					// If it's the final edge overall for that node, the final balance must be 0
+					if (RemainingInDegree(edge.From) == 0 && Balance(edge.From) - (long)edge.Value != 0)
+					{
+						throw new InvalidOperationException("Can't add final in edge without discharging negative value completely.");
+					}
+				}
+			}
+
+			return this with
+			{
+				Predecessors = Predecessors.SetItem(edge.To, predecessors.Add(edge)),
+				Successors = Successors.SetItem(edge.From, successors.Add(edge)),
+				EdgeBalances = EdgeBalances.SetItems(
+					new KeyValuePair<RequestNode, long>[]
+					{
+						new (edge.From, EdgeBalances[edge.From] - (long)edge.Value),
+						new (edge.To,   EdgeBalances[edge.To]   + (long)edge.Value),
+					}),
+			};
+		}
+
+		// Find the largest negative or positive balance node for the given
+		// credential type, and one or more smaller nodes with a combined total
+		// magnitude exceeding that of the largest magnitude node when possible.
+		public (RequestNode largestMagnitudeNode, IEnumerable<RequestNode> smallMagnitudeNodes, bool fanIn) MatchNodesToDischarge(IEnumerable<RequestNode> nodesWithRemainingOutDegree, IEnumerable<RequestNode> nodesWithRemainingInDegree)
+		{
+			ImmutableArray<RequestNode> sources = nodesWithRemainingOutDegree
+				.OrderByDescending(v => Balance(v))
+				.ThenByDescending(v => RemainingOutDegree(v))
+				.ThenByDescending(v => AvailableZeroOutDegree(v))
+				.ToImmutableArray();
+
+			ImmutableArray<RequestNode> sinks = nodesWithRemainingInDegree
+				.OrderBy(v => Balance(v))
+				.ThenByDescending(v => RemainingInDegree(v))
+				.ToImmutableArray();
+
+			var nSources = 1;
+			var nSinks = 1;
+
+			long SourcesSum() => sources.Take(nSources).Sum(v => Balance(v));
+			long SinksSum() => sinks.Take(nSinks).Sum(v => Balance(v));
+			long CompareSums() => SourcesSum().CompareTo(-1 * SinksSum());
+
+			// We want to fully discharge the larger (in absolute magnitude) of
+			// the two nodes, so we will add more nodes to the smaller one until
+			// we can fully cover. At each step of the iteration we fully
+			// discharge at least 2 nodes from the queue.
+			var initialComparison = CompareSums();
+			var fanIn = initialComparison == -1;
+
+			if (initialComparison != 0 && SinksSum() != 0)
+			{
+				Action takeOneMore = fanIn ? () => nSources++ : () => nSinks++;
+
+				// Take more nodes until the comparison sign changes or
+				// we run out.
+				while (initialComparison == CompareSums()
+					   && (fanIn ? sources.Length - nSources > 0
+								 : sinks.Length - nSinks > 0))
+				{
+					takeOneMore();
+				}
+			}
+
```

Comment:
It is really messy. I improved it here: https://github.com/lontivero/Wasabito/blob/8332cfc948d10c09bf4f823180970b9323b0669b/WalletWasabi/WabiSabi/Client/CredentialDependencies/CredentialEdgeSet.cs#L115-L134

In case it is not correct or doesn't fit well here I hope it inspires you.

Reply:
Yes, but I tried to work with your approach and found it harder to reason about this one, i'm still trying to find an in between point. Specifically the point of using these nested functions is to avoid passing variables to them, that's why I added that in the first place, your version retains that nested complexity but still passes around explicit arguments which I found confusing.

I think the right middle ground is to have your more recursive control flow but try and structure it more conventionally, but since this was a self contained method and I felt kind of stuck improving its readability I just let it go in the last round of changes